transactions/transactor.go

```diff
@@ -165,6 +145,82 @@ func (t *Transactor) SendTransactionWithSignature(args SendTxArgs, sig []byte) (
 	return signedTx.Hash(), nil
 }
 
+func (t *Transactor) HashTransaction(args SendTxArgs) (validatedArgs SendTxArgs, hash gethcommon.Hash, err error) {
+	if !args.Valid() {
+		return validatedArgs, hash, ErrInvalidSendTxArgs
+	}
+
+	validatedArgs = args
+
+	t.addrLock.LockAddr(args.From)
+	defer func() {
+		t.addrLock.UnlockAddr(args.From)
+	}()
+
+	var localNonce uint64
+	if val, ok := t.localNonce.Load(args.From); ok {
+		localNonce = val.(uint64)
+	}
+
+	var nonce uint64
```

Comment:
I would extract:
```go
func (t *Transactor) getTxNonce() {
    var (
        localNonce uint64
        nonce uint64
    )

    if val, ok := t.localNonce.Load(args.From); ok {
        localNonce = val.(uint64)
    }

    ctx, cancel := context.WithTimeout(context.Background(), t.rpcCallTimeout)
    defer cancel()
    nonce, err = t.pendingNonceProvider.PendingNonceAt(ctx, args.From)
    if err != nil {
        return validatedArgs, hash, err
    }
    // if upstream node returned nonce higher than ours we will use it, as it probably means
    // that another client was used for sending transactions
    if localNonce > nonce {
        nonce = localNonce
    }

    return nonce
}
```
Interesting that we don't use nonce from `SendTxArgs`.

Getting gas price and gas can also be extracted to separate method. They look the same here and in `validateAndPropagate`.

Reply:
I'll refactor it, I'm not sure if we want to touch the other methods right now since they are used from SendTransaction though.

Reply:
We can do it in another PR but then let's create a GH issue.